### PR TITLE
Updating the emitter to more generally handle 4-Byte SSE4 instructions.

### PR DESCRIPTION
**Pull request by tannergooding**
FYI. @CarolEidt, @fiigii, @eerhardt

This should mostly resolve https://github.com/dotnet/coreclr/issues/15908 and https://github.com/dotnet/coreclr/issues/16216, at least for the code paths currently being executed.

### CODE REVIEW

**Comment by tannergooding** [comment]
There is also still an issue that needs resolving in the `IF_RRW_RRW_CNS`, which had an existing emit pattern different from the rest of the code.

**Comment by tannergooding** [comment]
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

**Comment by tannergooding** [comment]
test Windows_NT x64 Checked jitincompletehwintrinsic
test Windows_NT x64 Checked jitx86hwintrinsicnoavx
test Windows_NT x64 Checked jitx86hwintrinsicnoavx2
test Windows_NT x64 Checked jitx86hwintrinsicnosimd
test Windows_NT x64 Checked jitnox86hwintrinsic

test Windows_NT x86 Checked jitincompletehwintrinsic
test Windows_NT x86 Checked jitx86hwintrinsicnoavx
test Windows_NT x86 Checked jitx86hwintrinsicnoavx2
test Windows_NT x86 Checked jitx86hwintrinsicnosimd
test Windows_NT x86 Checked jitnox86hwintrinsic

test Ubuntu x64 Checked jitincompletehwintrinsic
test Ubuntu x64 Checked jitx86hwintrinsicnoavx
test Ubuntu x64 Checked jitx86hwintrinsicnoavx2
test Ubuntu x64 Checked jitx86hwintrinsicnosimd
test Ubuntu x64 Checked jitnox86hwintrinsic

test OSX10.12 x64 Checked jitincompletehwintrinsic
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx
test OSX10.12 x64 Checked jitx86hwintrinsicnoavx2
test OSX10.12 x64 Checked jitx86hwintrinsicnosimd
test OSX10.12 x64 Checked jitnox86hwintrinsic

File: src/jit/emitxarch.cpp [hunk 1]
```diff
@@ -8420,21 +8453,24 @@ BYTE* emitter::emitOutputAM(BYTE* dst, instrDesc* id, code_t code, CnsVal* addc)
...
+        {
+            regNumber src1 = id->idReg2();
+
+            if ((id->idInsFmt() != IF_RWR_RRD_ARD) && (id->idInsFmt() != IF_RWR_RRD_ARD_CNS))
+            {
+                src1 = id->idReg1();
+            }
 
-        if ((id->idInsFmt() != IF_RWR_RRD_ARD) && (id->idInsFmt() != IF_RWR_RRD_ARD_CNS))
+            // encode source operand reg in 'vvvv' bits in 1's compliement form
```
Comment (CarolEidt) [inline src/jit/emitxarch.cpp]: While you're here you could fix the typo: "compliement" should be "complement"

File: src/jit/emitxarch.cpp [hunk 2]
```diff
@@ -10323,7 +10389,7 @@ BYTE* emitter::emitOutputRR(BYTE* dst, instrDesc* id)
         assert(IsAVXInstruction(ins) || IsSSE4Instruction(ins));
         if ((code & 0xFF00) == 0xC000)
         {
-            dst += emitOutputByte(dst, (0xC0 | regCode));
+            dst += emitOutputWord(dst, code | (regCode << 8));
```
Comment (CarolEidt) [inline src/jit/emitxarch.cpp]: There's probably something I'm missing here, but how did this go from emitting a byte to emitting a word?

Comment (tannergooding) [reply src/jit/emitxarch.cpp]: From what I can tell based on the comment above, this was trying to support the smaller encoding (which isn't supported anywhere else in the emitter).

I thought I had a comment indicating I was still looking at this in particular, but can't seem to find it anymore (maybe I just forget to submit).

Comment (tannergooding) [reply src/jit/emitxarch.cpp]: @CarolEidt, I'm finishing validating locally, but it looks like this is a dead code path and is not currently hit (for `emitOutputRR`). I can only speculate (based on the comment above this if block) that it was meant to support the smaller 2-byte VEX prefix scenario, which isn't actually working.

However, for `emitOutputInstr` it special cases `IF_RRW_RRW_CNS`, that does hit the equivalent code path, and that requires it to `emitOutputWord`.

I believe the correct fix is to refactor all three cases where this particular code pattern is (`emitOutputRR`, `emitOutputRRR`, and `IF_RRW_RRW_CNS` in `emitOutputInstr`) to do the following:
```C++
// TODO-XArch-CQ: Right now support 4-byte opcode instructions only
if ((code & 0xFF00) == 0xC000)
{
    dst += emitOutputWord(dst, code | (regCode << 8));
}
else if ((code & 0xFF) == 0x00)
{
    // This case happens for SSE4/AVX instructions only
    assert(IsAVXInstruction(ins) || IsSSE4Instruction(ins));

    dst += emitOutputByte(dst, (code >> 8) & 0xFF);
    dst += emitOutputByte(dst, (0xC0 | regCode));
}
else
{
    dst += emitOutputWord(dst, code);
    dst += emitOutputByte(dst, (0xC0 | regcode));
}
```

Comment (tannergooding) [reply src/jit/emitxarch.cpp]: Changed to the above. I validated that the original code path was never hit for `emitOutputRR` and `emitOutputRRR`.

File: src/jit/emitxarch.cpp (hunk 1 above)
Comment (tannergooding) [reply src/jit/emitxarch.cpp]: I fixed it here, and the 12 other places it occurred in this document.

**Comment by tannergooding** [comment]
@dotnet-bot test Windows_NT x64 Checked jitstress1 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStress=1)
@dotnet-bot test Windows_NT x64 Checked jitstress2_jitstressregs0x1000 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStress=2 JitStressRegs=0x1000)
@dotnet-bot test Windows_NT x64 Checked jitstress2_jitstressregs0x10 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStress=2 JitStressRegs=0x10)
@dotnet-bot test Windows_NT x64 Checked jitstress2_jitstressregs0x80 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStress=2 JitStressRegs=0x80)
@dotnet-bot test Windows_NT x64 Checked jitstress2_jitstressregs1 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStress=2 JitStressRegs=1)
@dotnet-bot test Windows_NT x64 Checked jitstress2_jitstressregs2 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStress=2 JitStressRegs=2)
@dotnet-bot test Windows_NT x64 Checked jitstress2_jitstressregs3 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStress=2 JitStressRegs=3)
@dotnet-bot test Windows_NT x64 Checked jitstress2_jitstressregs4 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStress=2 JitStressRegs=4)
@dotnet-bot test Windows_NT x64 Checked jitstress2_jitstressregs8 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStress=2 JitStressRegs=8)
@dotnet-bot test Windows_NT x64 Checked jitstress2 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStress=2)
@dotnet-bot test Windows_NT x64 Checked jitstressregs0x1000 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStressRegs=0x1000)
@dotnet-bot test Windows_NT x64 Checked jitstressregs0x10 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStressRegs=0x10)
@dotnet-bot test Windows_NT x64 Checked jitstressregs0x80 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStressRegs=0x80)
@dotnet-bot test Windows_NT x64 Checked jitstressregs1 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStressRegs=1)
@dotnet-bot test Windows_NT x64 Checked jitstressregs2 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStressRegs=2)
@dotnet-bot test Windows_NT x64 Checked jitstressregs3 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStressRegs=3)
@dotnet-bot test Windows_NT x64 Checked jitstressregs4 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStressRegs=4)
@dotnet-bot test Windows_NT x64 Checked jitstressregs8 | Queues Windows_NT x64 Checked Build and Test (Jit -  JitStressRegs=8)

**Comment by tannergooding** [comment]
The following are a separate issue, tracked by https://github.com/dotnet/coreclr/issues/16236:
* x64_checked_windows_nt_jitstress2
* x64_checked_windows_nt_jitstress2_jitstressregs8
* x64_checked_windows_nt_jitstress2_jitstressregs4
* x64_checked_windows_nt_jitstress2_jitstressregs3
* x64_checked_windows_nt_jitstress2_jitstressregs2
* x64_checked_windows_nt_jitstress2_jitstressregs1
* x64_checked_windows_nt_jitstress2_jitstressregs0x80
* x64_checked_windows_nt_jitstress1

The following timed out and have been reset:
* x64_checked_windows_nt_jitstress2_jitstressregs0x1000
* x64_checked_windows_nt_jitstress2_jitstressregs0x10
* x64_checked_windows_nt_jitincompletehwintrinsic

The following looks new, might be related, investigating:
* x64_checked_windows_nt_jitstressregs4

The following has been reset, tests failed due to OOM, looks unrelated:
* x64_checked_windows_nt_jitx86hwintrinsicnoavx

File: src/jit/emitxarch.cpp (hunk 2 above)
Comment (CarolEidt) [reply src/jit/emitxarch.cpp]: Awesome! Thanks for checking this out and cleaning it up.

**Comment by tannergooding** [comment]
x64_checked_windows_nt_jitstressregs4 is not related. The `AddRex*Prefix` checks need to be updated to account for the prefetch instructions (which are SSE instructions, but which need the actual REX prefix, rather than the VEX prefix).

I'll log a bug and have a fix up tonight.

**Comment by 4creators** [comment]
> The following has been reset, tests failed due to OOM, looks unrelated:

seems to be related to `Test Infrastructure Failure: The paging file is too small for this operation to complete.` failures in #16237

**Pull request merged**